src/msg/node.h

```diff
@@ -39,25 +40,49 @@ class CRNode {
 
     void SetRunner(std::shared_ptr<JobRunner> runner);
 
-    bool AddJobToRunner(job_t&& job);
+    bool AddJobToRunner(job_t&& job, strand_ptr_t strand);
+
+    bool AddJobToRunner(job_t&& job) { return AddJobToRunner(std::move(job), nullptr); }
 
     bool AddMessageToRunner(const CRMessageBasePtr& message);
 
-    // Not thread-safe, do not call concurrently nor call it when messages are coming,
+    // Subscribing functions are NOT thread-safe, do not call concurrently nor call it when messages are coming,
```

Comment:
Wondering if you can also leave a data race for tsan to pick up.

Reply:
Doesn't have to do it in this PR, maybe we can come up with some systematic way for this kind of flag.

Reply:
Also, why not just make it thread safe?

Reply:
I'm not saying we need to make it TS, but to say you can:
1. Create SIMPLE data race for TSAN to pick up if misused.
2. Add detector, e.g. a RAII atomi<bool> set when inside function and checked to log error if already set.

Reply:
> Create SIMPLE data race for TSAN to pick up if misused.

It accesses a shared map. So TSAN will catch it if there is concurrency.

> Add detector, e.g. a RAII atomic set when inside function and checked to log error if already set.

Any suggestions? There should not be concurrent {,un}subscriptions, and no new {,un}subscriptions should happen if there are messages flying in the system.

Reply:
If you have any good ideas, feel free to add in the following PR.

Reply:
> If you have any good ideas, feel free to add in the following PR.

As said, a lockfree lock, implemented by atomic bool, to emit warning on contention.
Basically an unique_lock with try_lock + check ownership, but without false negative.

Reply:
Will handle that in the following PR.